worker_test.go

```diff
@@ -2,21 +2,380 @@ package main
 
 import (
 	"context"
+	"crypto/rand"
+	"crypto/rsa"
 	"crypto/x509"
+	"crypto/x509/pkix"
+	"fmt"
+	"math/big"
+	"net/http/httptest"
+	"os"
 	"reflect"
 	"sync"
 	"testing"
 	"time"
 
+	"github.com/aws/aws-sdk-go/service/autoscaling"
 	cloudformation "github.com/mweagle/go-cloudformation"
+	log "github.com/sirupsen/logrus"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
-	"github.com/zalando-incubator/kube-ingress-aws-controller/aws"
+	awsAdapter "github.com/zalando-incubator/kube-ingress-aws-controller/aws"
 	"github.com/zalando-incubator/kube-ingress-aws-controller/certs"
 	"github.com/zalando-incubator/kube-ingress-aws-controller/kubernetes"
+	"github.com/zalando/skipper/dataclients/kubernetes/kubernetestest"
 	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/zalando-incubator/kube-ingress-aws-controller/aws/fake"
 )
 
+// TODO(LThiesen): This should be extracted to another file.
+// This is also just a copy of certs test so it might be useful
+// to just distribute it in a separate package like aws/fake
+type mockedCertificateProvider struct {
+	t *testing.T
+}
+
+func (m mockedCertificateProvider) GetCertificates() ([]*certs.CertificateSummary, error) {
+	//tenYears := time.Hour * 24 * 365 * 10
+	/*caCert := x509.Certificate{
+		SerialNumber: big.NewInt(123),
+		Subject: pkix.Name{
+			Organization: []string{"Testing CA"},
+			CommonName:   "foo.bar.org",
+		},
+		NotBefore: time.Time{},
+		NotAfter:  time.Now().Add(tenYears),
+
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+
+		DNSNames: []string{"foo.bar.org"},
+		IsCA:     false,
+	}*/
+
+	/*caCert := x509.Certificate{
+		SerialNumber: big.NewInt(123),
+		DNSNames:     []string{"foo.bar.org"},
+		NotBefore:    time.Now(),
+		NotAfter:     time.Now().Add(time.Hour * 24),
+
+		KeyUsage:              x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		BasicConstraintsValid: true,
+
+		IsCA: true,
+	}*/
+
+	certSummary := createDummyCertDetail(m.t, "DUMMY", []string{"foo.bar.org"}, time.Now(), time.Now().Add(time.Hour*24))
+
+	/*certificateSummary := certs.NewCertificate(
+	"foo.bar.org",
+	&caCert,
+	nil)*/
+	return []*certs.CertificateSummary{certSummary}, nil
+}
+
+type caInfra struct {
+	sync.Once
```

Comment:
Hmm, I'm not sure how to fix this. We are embedding the type `sync.Once` within the struct `caInfra`. What I can see as an option here is to give the struct a better name, indicating it is a singleton or something like that. Do you think this would address your concern?

Reply:
Oh, you are suggesting to not embedd this type within the struct, but instead have a `once sync.Once` field and use that instead.

Reply:
I agree with you on the readability side but there are two problems. With a explicit once declaration whenever we create a new `caInfra` a copy of once is created making it actually not lock the ca initialization. The other problem is that the linter does not allow me. So I guess I'll just make explicit in the type name that this is a singleton, which I think is the better I can do to improve readability in this case.